src/entries/background/handlers/handleProviderRequest.ts

```diff
@@ -66,62 +67,106 @@ const focusOnWindow = (windowId: number) => {
 const openWindowForTabId = async (tabId: string) => {
   const { notificationWindows } = useNotificationWindowStore.getState();
   const notificationWindow = notificationWindows[tabId];
-  if (notificationWindow) {
-    chrome.windows.get(
-      notificationWindow.id as number,
-      async (existingWindow) => {
-        if (chrome.runtime.lastError) {
-          createNewWindow(tabId);
+  if (notificationWindow?.id) {
+    chrome.windows.get(notificationWindow.id, async (existingWindow) => {
+      if (chrome.runtime.lastError) {
+        createNewWindow(tabId);
+      } else {
+        if (existingWindow.id) {
+          focusOnWindow(existingWindow.id);
         } else {
-          if (existingWindow) {
-            focusOnWindow(existingWindow.id as number);
-          } else {
-            createNewWindow(tabId);
-          }
+          createNewWindow(tabId);
         }
-      },
-    );
+      }
+    });
   } else {
     createNewWindow(tabId);
   }
 };
 
+const handleConnectionResolver = (
+  host: string,
+  addConnectionResolver: (
+    host: string,
+    resolver: {
+      resolve: (value: object) => void;
+      reject: (error: Error) => void;
+    },
+  ) => void,
+): Promise<object> => {
+  return new Promise((resolve, reject) => {
+    addConnectionResolver(host, { resolve, reject });
+  });
+};
+
+const resolveConnectionRequestsIfNeeded = (
```

Comment:
Is there any duplicative implementation between this concept and `handleTabAndWindowUpdates`? We attach to the window/tab listeners there for things like expiry